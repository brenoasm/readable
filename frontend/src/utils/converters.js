import moment from 'moment';

const convertDate = (value, format) => moment(value).format(format);

export {
  convertDate
};
