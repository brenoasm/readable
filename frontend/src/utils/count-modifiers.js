export const SUBTRACT = 'subtract';
export const ADD = 'add';
