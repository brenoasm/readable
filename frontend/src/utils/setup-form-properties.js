export const setupFormProperties = (formProperties) => {
  if (
    Object.keys(formProperties).length === 0 &&
    formProperties.constructor === Object
  )
  return {};

  let formattedObject = {
    formIsValid: false
  };

  Object.keys(formProperties).forEach(prop => {
    const param = formProperties[prop];

    try {
      let validations = [];
      let errors = [];
      const value = '';

      let objectToReturn = {
        ...param,
        value: param.value || value,
        isDirty: param.isDirty || false
      };

      if (param.validations) {
        validations = [
          ...param.validations.map(
            validation =>
            typeof validation !== 'function'
            ? Error('Invalid param. It should be a function')
            : validation
          )
          //criar funções padrões
        ];
      }

      formattedObject = {
        ...formattedObject,
        [prop]: {
          ...objectToReturn,
          errors,
          validations,
          name: prop
        }
      };
    } catch (e) {
      throw e;
    }
  });

  return formattedObject;
};
