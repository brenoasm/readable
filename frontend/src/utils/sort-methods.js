export const SORT_BY_DATE = 'sort_by_date';
export const SORT_BY_VOTE_SCORE = 'sort_by_vote_score';

export default [
  {
    text: 'Por data',
    value: SORT_BY_DATE
  },
  {
    text: 'Por votos',
    value: SORT_BY_VOTE_SCORE
  }
];
