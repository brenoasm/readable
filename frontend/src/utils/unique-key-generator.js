export const getRandomInt = (min, max) => {
  min = Math.ceil(min);
  max = Math.floor(max);
  return Math.floor(Math.random() * (max - min)) + min;
};

export const getRandomId = (min, max, timestamp) => {
  const randomInt = getRandomInt(min, max);

  const randomId = timestamp.split('').map(char => {
    const randomIntAux = getRandomInt(min, max);

    return `${randomInt}${char}${randomIntAux}`
  }).join('');

  return randomId;
};
