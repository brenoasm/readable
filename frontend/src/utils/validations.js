export const validValue = value => {
  if (!value) return 'Campo requirido';
};

export const validOption = value => {
  if (!value) return 'Selecione uma opção válida';
};

export const isTheFormValid = (properties, actualPropertyName, isValid) => {
  return Object.keys(properties)
    .map(
      prop =>
        properties[prop].isFormField
          ? prop === actualPropertyName
            ? isValid
            : properties[prop].isValid
          : true
    )
    .every(isThePropertyValid => isThePropertyValid);
};

export const runValidations = (validations, value) => {
  let errors = [];

  validations.forEach(validation => {
    const result = validation(value);

    if (result) errors.push(result);
  });

  return errors;
};
