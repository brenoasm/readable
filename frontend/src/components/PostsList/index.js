import React from 'react';
import styled from 'styled-components';
import * as _ from 'lodash';

import PostRow from '../PostRow';

import { SORT_BY_DATE } from '../../utils/sort-methods';

const StyledPostsList = styled.ul`
  display: flex;
  flex-direction: column;
  padding: 0;

  li {
    margin: 10px;
  }
`;

const PostsList = ({ posts, modifyVotes, sortMethod, editPost }) => {
  const orderedPosts =
    sortMethod === SORT_BY_DATE
      ? _.sortBy(posts, post => post.timestamp)
      : _.sortBy(posts, post => post.voteScore).reverse();

  return (
    <StyledPostsList>
      {Array.isArray(orderedPosts) &&
        orderedPosts.map(post => (
          <li>
            <PostRow modifyVotes={modifyVotes} key={post.id} post={post} editPost={editPost} />
          </li>
        ))}
    </StyledPostsList>
  );
};

export default PostsList;
