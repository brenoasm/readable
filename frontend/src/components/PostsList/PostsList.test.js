/* eslint-disable */

import React from 'react';

import PostsList from './';

import { SORT_BY_DATE, SORT_BY_VOTE_SCORE } from '../../utils/sort-methods';
import PostMock from '../../testHelpers/mocks/posts';

describe('PostsList', () => {
  const initialProps = {
    posts: [
      PostMock.posts[0],
      PostMock.posts[1]
    ],
    modifyPostVoteValues: jest.fn(),
    sortMethod: SORT_BY_DATE,
    editPost: jest.fn(),
    deletePost: jest.fn()
  };

  it('matches the snapshot', () => {
    const wrapper = shallow(<PostsList {...initialProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should order the posts by date desc', () => {
    const wrapper = shallow(<PostsList {...initialProps} />);

    const postsRow = wrapper.find('PostRow');

    expect(postsRow.get(0).props.post.timestamp)
      .toBeGreaterThan(postsRow.get(1).props.post.timestamp);
  });

  it('should order the posts by voteScore desc', () => {
    const alteredProps = {
      ...initialProps,
      sortMethod: SORT_BY_VOTE_SCORE,
    }

    const wrapper = shallow(<PostsList {...alteredProps} />);

    const postsRow = wrapper.find('PostRow');

    expect(postsRow.get(0).props.post.voteScore)
      .toBeGreaterThan(postsRow.get(1).props.post.voteScore);
  });
});
