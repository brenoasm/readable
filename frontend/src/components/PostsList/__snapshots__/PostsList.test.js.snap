// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostsList matches the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PostsList
    deletePost={[Function]}
    editPost={[Function]}
    modifyPostVoteValues={[Function]}
    posts={
        Array [
            Object {
              "author": "thingone",
              "body": "Just kidding. It takes more than 10 minutes to learn technology.",
              "category": "redux",
              "commentCount": 0,
              "deleted": false,
              "id": "1",
              "timestamp": 1468479767190,
              "title": "Learn Redux in 10 minutes!",
              "voteScore": 1,
            },
            Object {
              "author": "thingtwo",
              "body": "Everyone says so after all.",
              "category": "react",
              "commentCount": 1,
              "deleted": false,
              "id": "2",
              "timestamp": 1467166872634,
              "title": "Udacity is the best place to learn React",
              "voteScore": -1,
            },
          ]
    }
    sortMethod="sort_by_date"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <li>
          <PostRow
                    deletePost={[Function]}
                    editPost={[Function]}
                    modifyPostVoteValues={[Function]}
                    post={
                              Object {
                                        "author": "thingone",
                                        "body": "Just kidding. It takes more than 10 minutes to learn technology.",
                                        "category": "redux",
                                        "commentCount": 0,
                                        "deleted": false,
                                        "id": "1",
                                        "timestamp": 1468479767190,
                                        "title": "Learn Redux in 10 minutes!",
                                        "voteScore": 1,
                                      }
                    }
          />
</li>,
        <li>
          <PostRow
                    deletePost={[Function]}
                    editPost={[Function]}
                    modifyPostVoteValues={[Function]}
                    post={
                              Object {
                                        "author": "thingtwo",
                                        "body": "Everyone says so after all.",
                                        "category": "react",
                                        "commentCount": 1,
                                        "deleted": false,
                                        "id": "2",
                                        "timestamp": 1467166872634,
                                        "title": "Udacity is the best place to learn React",
                                        "voteScore": -1,
                                      }
                    }
          />
</li>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "1",
        "nodeType": "host",
        "props": Object {
          "children": <PostRow
            deletePost={[Function]}
            editPost={[Function]}
            modifyPostVoteValues={[Function]}
            post={
                        Object {
                                    "author": "thingone",
                                    "body": "Just kidding. It takes more than 10 minutes to learn technology.",
                                    "category": "redux",
                                    "commentCount": 0,
                                    "deleted": false,
                                    "id": "1",
                                    "timestamp": 1468479767190,
                                    "title": "Learn Redux in 10 minutes!",
                                    "voteScore": 1,
                                  }
            }
/>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "deletePost": [Function],
            "editPost": [Function],
            "modifyPostVoteValues": [Function],
            "post": Object {
              "author": "thingone",
              "body": "Just kidding. It takes more than 10 minutes to learn technology.",
              "category": "redux",
              "commentCount": 0,
              "deleted": false,
              "id": "1",
              "timestamp": 1468479767190,
              "title": "Learn Redux in 10 minutes!",
              "voteScore": 1,
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "li",
      },
      Object {
        "instance": null,
        "key": "2",
        "nodeType": "host",
        "props": Object {
          "children": <PostRow
            deletePost={[Function]}
            editPost={[Function]}
            modifyPostVoteValues={[Function]}
            post={
                        Object {
                                    "author": "thingtwo",
                                    "body": "Everyone says so after all.",
                                    "category": "react",
                                    "commentCount": 1,
                                    "deleted": false,
                                    "id": "2",
                                    "timestamp": 1467166872634,
                                    "title": "Udacity is the best place to learn React",
                                    "voteScore": -1,
                                  }
            }
/>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "deletePost": [Function],
            "editPost": [Function],
            "modifyPostVoteValues": [Function],
            "post": Object {
              "author": "thingtwo",
              "body": "Everyone says so after all.",
              "category": "react",
              "commentCount": 1,
              "deleted": false,
              "id": "2",
              "timestamp": 1467166872634,
              "title": "Udacity is the best place to learn React",
              "voteScore": -1,
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "li",
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <li>
            <PostRow
                        deletePost={[Function]}
                        editPost={[Function]}
                        modifyPostVoteValues={[Function]}
                        post={
                                    Object {
                                                "author": "thingone",
                                                "body": "Just kidding. It takes more than 10 minutes to learn technology.",
                                                "category": "redux",
                                                "commentCount": 0,
                                                "deleted": false,
                                                "id": "1",
                                                "timestamp": 1468479767190,
                                                "title": "Learn Redux in 10 minutes!",
                                                "voteScore": 1,
                                              }
                        }
            />
</li>,
          <li>
            <PostRow
                        deletePost={[Function]}
                        editPost={[Function]}
                        modifyPostVoteValues={[Function]}
                        post={
                                    Object {
                                                "author": "thingtwo",
                                                "body": "Everyone says so after all.",
                                                "category": "react",
                                                "commentCount": 1,
                                                "deleted": false,
                                                "id": "2",
                                                "timestamp": 1467166872634,
                                                "title": "Udacity is the best place to learn React",
                                                "voteScore": -1,
                                              }
                        }
            />
</li>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "1",
          "nodeType": "host",
          "props": Object {
            "children": <PostRow
              deletePost={[Function]}
              editPost={[Function]}
              modifyPostVoteValues={[Function]}
              post={
                            Object {
                                          "author": "thingone",
                                          "body": "Just kidding. It takes more than 10 minutes to learn technology.",
                                          "category": "redux",
                                          "commentCount": 0,
                                          "deleted": false,
                                          "id": "1",
                                          "timestamp": 1468479767190,
                                          "title": "Learn Redux in 10 minutes!",
                                          "voteScore": 1,
                                        }
              }
/>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "deletePost": [Function],
              "editPost": [Function],
              "modifyPostVoteValues": [Function],
              "post": Object {
                "author": "thingone",
                "body": "Just kidding. It takes more than 10 minutes to learn technology.",
                "category": "redux",
                "commentCount": 0,
                "deleted": false,
                "id": "1",
                "timestamp": 1468479767190,
                "title": "Learn Redux in 10 minutes!",
                "voteScore": 1,
              },
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "li",
        },
        Object {
          "instance": null,
          "key": "2",
          "nodeType": "host",
          "props": Object {
            "children": <PostRow
              deletePost={[Function]}
              editPost={[Function]}
              modifyPostVoteValues={[Function]}
              post={
                            Object {
                                          "author": "thingtwo",
                                          "body": "Everyone says so after all.",
                                          "category": "react",
                                          "commentCount": 1,
                                          "deleted": false,
                                          "id": "2",
                                          "timestamp": 1467166872634,
                                          "title": "Udacity is the best place to learn React",
                                          "voteScore": -1,
                                        }
              }
/>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "deletePost": [Function],
              "editPost": [Function],
              "modifyPostVoteValues": [Function],
              "post": Object {
                "author": "thingtwo",
                "body": "Everyone says so after all.",
                "category": "react",
                "commentCount": 1,
                "deleted": false,
                "id": "2",
                "timestamp": 1467166872634,
                "title": "Udacity is the best place to learn React",
                "voteScore": -1,
              },
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "li",
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
